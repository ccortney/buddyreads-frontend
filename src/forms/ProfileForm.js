import { useState, useContext } from "react";
import {    Grid, 
            TextField, 
            Typography, 
            FormControl, 
            Button, 
            InputLabel, 
            OutlinedInput, 
            InputAdornment, 
            IconButton} from "@mui/material";
import VisibilityOff from '@mui/icons-material/VisibilityOff';
import Visibility from '@mui/icons-material/Visibility';
import UserContext from "../auth/UserContext";
import ErrorAlert from "../common/ErrorAlert";
import BuddyReadApi from "../api/api";

/** Profile editing form.
 *
 * Displays profile form and password change form. 
 * Handles changes to local form state.
 * Submitting the form calls the API to save, and triggers user reloading
 * throughout the site.
 *
 * Confirmation of a successful save is normally a simple <ErrorAlert>
 *
 * Routed as /profile
 * AppRoutes -> ProfileForm -> ErrorAlert
 */

const ProfileForm = ({editProfile}) => {
    const { currentUser, setCurrentUser } = useContext(UserContext)

    const INITIAL_STATE = {
        firstName: '',
        lastName: '',
        email: '', 
        password: '',
        showPassword: false
    }

    const [formData, setFormData] = useState(INITIAL_STATE);
    const [formErrors, setFormErrors] = useState([]);
    const [saveConfirmed, setSaveConfirmed] = useState({profile: false, password: false})

    const handleChange = (e) => {
        const {name, value} = e.target;
        setFormData(formData => ({...formData, [name]: value}))
        setFormErrors([]);
    }

    const handleSubmit = type => async e => {
        e.preventDefault();
        let updatedUser;

        let profileData = {
            email: formData.email || currentUser.email,
            password: formData.password || currentUser.password, 
            firstName: formData.firstName || currentUser.firstName,
            lastName: formData.lastName || currentUser.lastName, 
        }

        try {
            await editProfile(currentUser.id, profileData);
            updatedUser = await BuddyReadApi.getUser(currentUser.id)
        } catch(errors) {
            setFormErrors(errors);
            return;
        }

        setFormErrors([]);
        setCurrentUser(updatedUser);
        setFormData(INITIAL_STATE);
        setSaveConfirmed(data => ({...data, [type]: true}));
    }
      
    const handleClickShowPassword = () => {
        setFormData({
        ...formData,
        showPassword: !formData.showPassword,
        });
    };
      
    const handleMouseDownPassword = (event) => {
        event.preventDefault();
    };

  return (
        <Grid container align='center' sx={{justifyContent:'center'}}>
            <Grid item>
                {formErrors.length ? <ErrorAlert severity='error' messages={formErrors}/> : null }   
            </Grid>
          
            <Grid item>
                <form onSubmit={handleSubmit('profile')}>
                    <Typography gutterBottom variant="h5" component="div" sx={{pt: 3}}>
                        Edit Profile for {currentUser.firstName}
                    </Typography>
                    <Grid container alignItems="center" justifyContent="center" direction="column">
                        <Grid item >
                            <TextField
                                sx={{m: 1, width: 300}}
                                id="firstName"
                                name="firstName"
                                label="First Name"
                                type="text"
                                value={formData.firstName}
                                onChange={handleChange}
                            />
                        </Grid>
                        <Grid item>
                            <TextField
                                    sx={{m: 1, width: 300}}
                                    id="lastName"
                                    name="lastName"
                                    label="Last Name"
                                    type="text"
                                    value={formData.lastName}
                                    onChange={handleChange}
                                />
                        </Grid>
                        <Grid item>
                            <TextField
                                sx={{m: 1, width: 300}}
                                id="email"
                                name="email"
                                label="Email"
                                type="text"
                                value={formData.email}
                                onChange={handleChange}
                            />
                        </Grid>
                        <Button variant="contained" 
                                color="primary" 
                                type="submit" sx={{mt: 2}}>
                        Save Changes
                        </Button>
                        {saveConfirmed.profile ? <ErrorAlert severity='success' messages={["Profile Updated Successfully"]}/> : null } 
                    </Grid>
                </form>
                <form onSubmit={handleSubmit('password')}>
                <Typography gutterBottom variant="h5" component="div" sx={{pt: 3}}>
                    Edit Password
                </Typography>
                <Grid container alignItems="center" justifyContent="center" direction="column">

                    <Grid item >
                        <FormControl sx={{ m: 1, width: 300 }} variant="outlined">
                                <InputLabel htmlFor="password">New Password</InputLabel>
                                <OutlinedInput
                                id="password"
                                type={formData.showPassword ? 'text' : 'password'}
                                name="password"
                                value={formData.password}
                                onChange={handleChange}
                                endAdornment={
                                    <InputAdornment position="end">
                                    <IconButton
                                        aria-label="toggle password visibility"
                                        onClick={handleClickShowPassword}
                                        onMouseDown={handleMouseDownPassword}
                                        edge="end"
                                    >
                                    {formData.showPassword ? <VisibilityOff /> : <Visibility />}
                                    </IconButton>
                                    </InputAdornment>
                                }
                                label="Password"
                                />
                        </FormControl>   
                    </Grid>        
                    <Button variant="contained" 
                            color="primary" 
                            type="submit" 
                            sx={{mt: 2}} >
                        Save
                    </Button>
                    </Grid>
                </form>
                {saveConfirmed.password ? <ErrorAlert severity='success' messages={["Password Saved"]}/> : null } 
            </Grid>
        </Grid>
  );
};

export default ProfileForm;