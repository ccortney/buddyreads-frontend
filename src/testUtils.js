import UserContext from "./auth/UserContext";

const demoUser = {
  id: 0,
  first_name: "testfirst",
  last_name: "testlast",
  email: "test@test.net",
};

const UserProvider =
    ({ children, currentUser = demoUser }) => (
    <UserContext.Provider value={{ currentUser }}>
      {children}
    </UserContext.Provider>
);

export { UserProvider };
