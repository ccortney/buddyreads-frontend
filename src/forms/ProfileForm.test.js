import { render } from "@testing-library/react";
import ProfileForm from "./ProfileForm";
import { UserProvider } from "../testUtils";
import { MemoryRouter } from "react-router";

it("matches snapshot", function () {
  const { asFragment } = render(
      <MemoryRouter>
          <UserProvider>
            <ProfileForm />
          </UserProvider>
      </MemoryRouter>,
  );
  expect(asFragment()).toMatchSnapshot();
});
