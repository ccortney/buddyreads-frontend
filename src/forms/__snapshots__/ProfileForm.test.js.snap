// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    align="center"
    class="MuiGrid-root MuiGrid-container css-1lym95h-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    />
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <form>
        <div
          class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-xhhpbl-MuiTypography-root"
        >
          Edit Profile for 
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-l6hioc-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiTextField-root css-esk3re-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="firstName"
                id="firstName-label"
              >
                First Name
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="firstName"
                  name="firstName"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-1ftyaf0"
                  >
                    <span>
                      First Name
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiTextField-root css-esk3re-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="lastName"
                id="lastName-label"
              >
                Last Name
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="lastName"
                  name="lastName"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-1ftyaf0"
                  >
                    <span>
                      Last Name
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiTextField-root css-esk3re-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="email"
                id="email-label"
              >
                Email
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="email"
                  name="email"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-1ftyaf0"
                  >
                    <span>
                      Email
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-lll1vm-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            Save Changes
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </form>
      <form>
        <div
          class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-xhhpbl-MuiTypography-root"
        >
          Edit Password
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-l6hioc-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root css-2m9kme-MuiFormControl-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="password"
              >
                New Password
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                  id="password"
                  name="password"
                  type="password"
                  value=""
                />
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
                >
                  <button
                    aria-label="toggle password visibility"
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
                    tabindex="0"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="VisibilityIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
                      />
                    </svg>
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </div>
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-1ftyaf0"
                  >
                    <span>
                      Password
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-lll1vm-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            Save
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
